src/clio/main/Config.h

```diff
@@ -0,0 +1,159 @@
+#ifndef CLIO_CONFIG_H
+#define CLIO_CONFIG_H
+
+#include <boost/json.hpp>
+#include <iostream>
+#include <optional>
+#include <string>
+#include <unordered_set>
+
+struct ETLSourceConfig
+{
+    std::string ip;
+    std::string wsPort;
+    std::string grpcPort;
+    std::vector<std::string> cacheCommands;
+};
+
+enum class CacheLoadStyle { ASYNC, SYNC, NOT_AT_ALL };
+
+struct CacheConfig
+{
+    std::uint32_t numDiffs;
+    CacheLoadStyle cacheLoadStyle;
+};
+
+struct ServerConfig
+{
+    std::string ip;
+    std::uint32_t port;
+};
+
+struct DOSGuardConfig
+{
+    std::uint32_t maxFetches;
+    std::uint32_t sweepInterval;
+    std::unordered_set<std::string> whitelist;
+
+private:
+    static DOSGuardConfig
+    parseDOSGuardConfig(boost::json::value& config);
+};
+
+struct DatabaseConfig
+{
+    std::string type;
+
+    virtual ~DatabaseConfig(){};
+
+    static DatabaseConfig
+    parseDatabaseConfig(boost::json::value& cofig);
+};
+
+struct MockDatabaseConfig : public DatabaseConfig
+{
+    MockDatabaseConfig()
+    {
+        type = "mock";
+    }
+};
+
+struct CassandraConfig : public DatabaseConfig
+{
+    struct CassandraOptions
+    {
+        std::optional<std::string> secureConnectBundle;
+        std::optional<std::string> contactPoints;
+        std::optional<std::string> keyspace;
+        std::optional<std::string> username;
+        std::optional<std::string> password;
+        std::optional<std::string> certfile;
+        std::optional<std::uint32_t> port;
+        std::optional<std::uint32_t> ttl;
+        std::string tablePrefix;
+        std::uint32_t replicationFactor;
+        std::uint32_t syncInterval;
+        std::uint32_t maxRequestsOutstanding;
+        std::uint32_t threads;
+    };
+
+    CassandraOptions cassandra;
+
+    CassandraConfig(boost::json::value& config);
+};
+
+struct PostgresConfig : public DatabaseConfig
+{
+    struct PostgresOptions
+    {
+        bool experimental;
+        bool rememberIp;
+        std::string username;
+        std::string password;
+        std::string contactPoint;
+        std::string database;
+        std::uint32_t writeInterval;
+        std::uint32_t timeout;
+        std::uint32_t maxConnections;
+    };
+
+    PostgresOptions postgres;
+
+    PostgresConfig(boost::json::value& config);
+};
+
+boost::json::object
+parseConfig(std::string const& filename);
+
+class Config
+{
+    boost::json::object json_;
+    std::shared_ptr<DatabaseConfig> dbConfigStore;
+
+public:
+    DatabaseConfig& database;
```

Comment:
Why is this a reference?

Reply:
This has to be polymorphic to account for different types of database configs.